File: src/main/java/org/hisp/dhis/rules/models/RuleEvaluationResult.java
package org.hisp.dhis.rules.models;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.List;

public record RuleEvaluationResult(
        @Nonnull Rule rule,
        @Nonnull List<RuleEffect> ruleEffects,
        boolean evaluatedAs,
        boolean error)
{

    public static RuleEvaluationResult evaluatedResult(@Nonnull Rule rule, @Nonnull List<RuleEffect> ruleEffects) {
        return new RuleEvaluationResult( rule, ruleEffects, true, false );
    }

    public static RuleEvaluationResult notEvaluatedResult(@Nonnull Rule rule) {
        return new RuleEvaluationResult( rule, new ArrayList<>(), false, false );
    }

    public static RuleEvaluationResult errorRule( @Nonnull Rule rule, @Nonnull String errorMessage ) {
        ArrayList<RuleEffect> effects = new ArrayList<>();

        effects.add(new RuleEffect(rule.uid(), RuleActionError.create(errorMessage), errorMessage));
        return new RuleEvaluationResult( rule, effects, false, true );
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleEvaluationResult.kt
package org.hisp.dhis.rules.models

data class RuleEvaluationResult(
    val rule: Rule,
    val ruleEffects: List<RuleEffect>,
    val evaluatedAs: Boolean,
    val error: Boolean
) {

    fun rule(): Rule {
        return rule
    }

    fun ruleEffects(): List<RuleEffect> {
        return ruleEffects
    }

    fun evaluatedAs(): Boolean {
        return evaluatedAs
    }

    fun error(): Boolean {
        return error
    }

    companion object {
        fun evaluatedResult(rule: Rule, ruleEffects: List<RuleEffect>): RuleEvaluationResult {
            return RuleEvaluationResult(rule, ruleEffects, true, false)
        }

        fun notEvaluatedResult(rule: Rule): RuleEvaluationResult {
            return RuleEvaluationResult(rule, emptyList(), false, false)
        }

        fun errorRule(rule: Rule, errorMessage: String): RuleEvaluationResult {
            val effects = listOf(RuleEffect(rule.uid, RuleActionError.create(errorMessage), errorMessage))
            return RuleEvaluationResult(rule, effects, false, true)
        }
    }
}
